api/envoy/extensions/filters/http/ext_proc/v3/processing_mode.proto

```diff
@@ -68,6 +69,11 @@ message ProcessingMode {
     // chunk. If the body exceeds the configured buffer limit, then the body contents
     // up to the buffer limit will be sent.
     BUFFERED_PARTIAL = 3;
+
+    // [#not-implemented-hide:]
+    // Stream the body to the server in pieces as they arrive at the proxy regardless
+    // the header processing state.
+    STREAMED_BEFORE_HEADER_RESP = 4;
```

Comment:
How will we model the combination of this and M:N as different body modes?

Reply:
We can define two body modes as well:
STREAMED_MxN
STREAMED_MxN_BEFORE_HEADER_RESP

With this, we can contain these kind body modes definition inside the BodySendMode enum, instead of keep adding new knobs in the top lever filter configuration.  I feel this makes it easily extendable.  Also, we can have different body modes for INBOUND or OUTBOUND traffic. If it is filter configuration, it will be applied to both directions.  Please correct me if I am wrong.

Reply:
Do we need to have separate MxN mode? Can we make this mode imply MxN? I see we keep relaxing the original very rigid state machine bit by bit. Why not just go all the way and implement mode with completely independent bidi streaming between ext_proc server and Envoy?

Reply:
It's a good idea to me to not define two modes for MxN, i.e, just have one STREAMED_MxN mode, which implies no waiting for header response.

However, I would consider still keeping three STREAMED modes:
1) STREAMED  ------ legacy streaming mode for existing customer
2) STREAMED_NO_HEADER_WAIT  ----- 1:1 streaming mode not waiting for header response.
3) STREAMED_MxN               ------- MxN streaming mode not waiting for header response

The benefit of separating 2) from 3) is that Envoy can still do security validation with mode 2), and the server do not have to send back the original body. This is very different from 3).

Reply:
Can we have 2 above (STREAMED_NO_HEADER_WAIT) as a config knob instead of a new STREAM mode? 

Technically, this PR should work for both 1:1 (need some changes though) and M:N. I am not very sure that configuring two different modes for the request and response paths is a very popular model.  However, if we really want, it can still be supported by a single filter config knob, for example a enum for request and response, respectively.

Reply:
For STREAMED_MxN mode, the body sending should be irrespective to whether header response  is received or not, i.e, Envoy should just send and not depend on a filter configuration to make choice.  This is what @yanavlasov is proposing. I agree with that.

Ideally, Envoy should be able to send body in 1:1 STREAMED mode irrespective to whether header response  is received or not as well. However, if we directly remove that check: https://github.com/envoyproxy/envoy/blob/0290ac0c30860bde16d3818cada367a216cc8de3/source/extensions/filters/http/ext_proc/ext_proc.cc#L492, that will cause backward compatibility issue. So let's only do that with the new STREAMED_NO_HEADER_WAIT configuration.

For your suggestion to add a separate filter config knob, that's what I had tried initially.  But I am seeing a couple of issues:

1) The body mode is scattered into two different places, making the logic a little bit complicated.
2) If the knob is a boolean, then it does not have the granularity to differentiate INBOUND or OUTBOUND direction.  If it is a enum as you suggested {request, response}, then why not adding that body modes in the BodySendMode enum, which is naturally can be configured into INBOUND or OUTBOUND direction to make it easy to understand.
3) On top of that, the MxN STREAMED mode should not depend on this filter configuration knob as mentioned above.

Reply:
My comment is mostly towards 1:1 and I don't feel any of issues you mentioned above could be the deal breaker for with enum filter config which also address your initial question on supporting inbound vs outbound

Personally I would suggest avoid growing the processing mode unless there is significant difference that we must have to do so.

Reply:
Yeah, from functional perspective, we can make it work with either way of the API changes. It's just from API perspective,  I am thinking contain the body modes inside the BodySendMode enum is cleaner and more extendable than got it from two different places.  
This probably has a lot of personal preferences. Let's consult our API experts @adisuissa @htuch , and we can do whatever they recommend.

Reply:
This thread is now quite long, can you summarize what the choices are?

Reply:
The choices are between these two:

1) Adding a filter top level configuration knob under "message ExternalProcessor":
      bool continue_body_streaming_without_header_resp = 21;


2)  Adding a body processing mode under " enum BodySendMode " 
   STREAMED_NO_HEADER_WAIT = 4;

Reply:
I think the question is whether there are situations (realistic ones) in which we don't know ahead-of-time whether async header behavior is desired. E.g. the same ext_proc server / configuration might want to dynamically switch based on the contents of a header. I'm not sure if I can spot an example, but interested to hear what others say.

Reply:
If the server do not know ahead of time, I think they can just operate on a less restricted body mode.

That being said, with the below comments about the processing_mode_overrides (https://github.com/envoyproxy/envoy/blob/6592faae7203ecba8d00ad6e965f9f1be44aea7a/api/envoy/service/ext_proc/v3/external_processor.proto#L184), it feels it is easier to keep this as a filter config knob.